finance/mip_raw_elt_batch.py

```diff
@@ -0,0 +1,285 @@
+            "filter[sessionPostedDate][gt]": max_session_posted_date,
+        },
+    )
+    total_records = initial_ledgers.json()["totalRecords"]
+    while len(all_ledgers) < total_records:
+        ledgers = requests.get(
+            _ledger_api,
+            headers={
+                "Authorization-Token": access_token,
+            },
+            params={
+                "page[number]": str(page_number),
+                "page[size]": str(page_size),
+                "filter[sessionPostedDate][gt]": max_session_posted_date,
+            },
+        )
+        page_number += 1
+        all_ledgers.extend(ledgers.json()["data"])
+    return all_ledgers
+
+
+def get_chart_of_accounts(
+    access_token: str, page_number: int = 0, page_size: int = 20
+) -> list:
+    """Get chart of accounts from the MIP API. This includes all the program codes
+
+    Args:
+        access_token (str): MIP access token
+        page_number (int, optional): page number. Defaults to 0.
+        page_size (int, optional): page size. Defaults to 20.
+
+    Returns:
+        list: List of charts of accounts
+    """
+    all_ledgers = []
+    initial_ledgers = requests.get(
+        _chart_of_accounts_api,
+        headers={
+            "Authorization-Token": access_token,
+        },
+        params={"page[number]": str(page_number), "page[size]": str(page_size)},
+    )
+    total_records = initial_ledgers.json()["total"]
+    while len(all_ledgers) < total_records:
+        ledgers = requests.get(
+            _chart_of_accounts_api,
+            headers={
+                "Authorization-Token": access_token,
+            },
+            params={"page[number]": str(page_number), "page[size]": str(page_size)},
+        )
+        page_number += 1
+        all_ledgers.extend(ledgers.json()["data"])
+    return all_ledgers
+
+
+def update_mip_raw_tables():
+    """Update MIP raw tables including the general ledger and chart of accounts"""
+    access_token = None
+    mips_creds = {
+        "username": "itops",
+        "password": MIP_AUTH,
+        "org": "SAGE_24146",
+    }
+    access_token = _request_login(mips_creds)
+    ctx = sc.connect(**conn_params)
+    cs = ctx.cursor()
+    results = cs.execute(
+        "SELECT MAX(SESSIONPOSTEDDATE) as max_session_posted_date FROM finance.mip_raw.ledgers"
+    )
+    max_session_posted_date = results.fetchone()[0]
+    # df = results.fetch_pandas_all()
+    formatted_dt = max_session_posted_date.strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]
+
+    ledgers = get_ledgers(
+        access_token, max_session_posted_date=formatted_dt, page_size=1000
+    )
+    if len(ledgers) > 0:
+        ledgers_df = pd.DataFrame(ledgers)
+        ledgers_df.to_csv("ledgers.csv", index=False)
+
+        results = cs.execute("SELECT * FROM finance.mip_raw.ledgers limit 1")
+        df = results.fetch_pandas_all()
+
+        ledgers_df.columns = ledgers_df.columns.str.upper()
+
+        for col, dtype in df.dtypes.items():
```

Comment:
You duplicated a whole block of casting logic between ledgers and chart of accounts.
 It’s DRY: Don’t Repeat Yourself. Not WET: Write Everything Twice.
:scissors: “Copy, paste, repeat. Who needs functions when you have caffeine and confidence?”